### PR TITLE
Migrate banned competitors to groups

### PR DESCRIPTION
**Pull request by danieljames-dj**
This PR is migration of banned competitors to groups.

This PR is ready for review for existing changes, but there are couple of more changes which might depend on other upcoming PRs:
- Update banned competitors is not working now
- Currently banned competitors list is shown only in leader panel, it has to be shown in some other relevant panels as well
I'm working on the above couple of issues and will submit them soon.
- Add to role change mailers

### CODE REVIEW

File: app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  const [formValues, setFormValues] = useState({
+    user: null,
+    endDate: banActionRole?.end_date,
+  });
+  const { save, saving } = useSaveAction();
+
+  const handleFormChange = (_, { name, value }) => setFormValues({ ...formValues, [name]: value });
+
+  const createNewBannedCompetitor = () => {
+    // Do action to create new banned competitor
```
Comment (kr-matthews) [inline app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx:23]: I don't think this comment is helpful, it just repeats the function name.
```suggestion
```

File: app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      groupType: groupTypes.banned_competitors,
+      endDate: formValues?.endDate,
+    }, () => {
+      sync();
+      closeForm();
+    }, { method: 'POST' });
+  };
+
+  const editBannedCompetitor = () => {
+    // Do action to edit new banned competitor
```
Comment (kr-matthews) [inline app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx:35]: Likewise. (Plus the word "new" shouldn't be there?)
```suggestion
```

File: app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }, { method: 'PATCH' });
+  };
+
+  const formSubmitHandler = () => {
+    if (banAction === 'new') {
+      createNewBannedCompetitor();
+    } else {
+      editBannedCompetitor();
+    }
+  };
```
Comment (kr-matthews) [inline app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx:50]: Does this work?
```suggestion
  const formSubmitHandler = banAction === 'new' ? createNewBannedCompetitor : editBannedCompetitor;
```

File: app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        showYearDropdown
+        dateFormatOverride="YYYY-MM-dd"
+        dropdownMode="select"
+        isoDate={formValues?.endDate}
+        onChange={(date) => handleFormChange(null, {
+          name: 'endDate',
+          value: date,
+        })}
+      />
+      <Form.Button onClick={() => closeForm()}>Cancel</Form.Button>
```
Comment (kr-matthews) [inline app/webpacker/components/Panel/pages/BannedCompetitorsPage/BannedCompetitorForm.jsx:80]:
```suggestion
      <Form.Button onClick={closeForm}>Cancel</Form.Button>
```

File: app/controllers/api/v0/user_roles_controller.rb [hunk 5]
```diff
@@ -39,7 +39,7 @@ class Api::V0::UserRolesController < Api::V0::ApiController
   # Filter role based on the permissions of the current user.
   private def can_current_user_access(role)
     group = role.group
-    !group.is_hidden || current_user&.has_permission?(:can_edit_groups, group.id)
+    !group.is_hidden || current_user&.has_permission?(:can_read_groups, group.id)
```
Comment (gregorbg) [inline app/controllers/api/v0/user_roles_controller.rb:42]: Without any context, this seems like a pretty crazy change: What does "access" (from the method name `can_user_access`) mean in this context?
And why is it suddenly enough to have read-only permissions when previously you needed write permissions?

File: app/controllers/api/v0/user_roles_controller.rb [hunk 6]
```diff
@@ -330,7 +335,8 @@ def update
       else
         return render status: :unprocessable_entity, json: { error: "Invalid parameter to be changed" }
       end
-    else
+    # elsif group_type == UserGroup.group_types[:banned_competitors]
+    # else
```
Comment (gregorbg) [inline app/controllers/api/v0/user_roles_controller.rb:339]: Why is there commented-out (dead) code here?

File: app/models/user.rb [hunk 7]
```diff
@@ -570,6 +570,17 @@ def banned_at_date?(date)
     delegate_roles.select { |role| role.metadata.status == RolesMetadataDelegateRegions.statuses[:senior_delegate] }
   end
 
+  private def groups_with_read_access
+    return "*" if can_edit_any_groups?
+    groups = groups_with_edit_access
+
+    if can_view_banned_competitors?
+      groups += UserGroup.banned_competitors.ids
+    end
```
Comment (gregorbg) [inline app/models/user.rb:579]: Why is this case not covered by `groups_with_edit_access` directly above?

File: db/schema.rb [hunk 8]
```diff
@@ -1141,6 +1148,7 @@
     t.datetime "created_at", precision: nil, null: false
     t.datetime "updated_at", precision: nil, null: false
     t.boolean "team_senior_member", default: false, null: false
+    t.bigint "group_id"
```
Comment (gregorbg) [inline db/schema.rb:1151]: I'm not sure but this feels unrelated?

File: db/schema.rb [hunk 9]
```diff
@@ -1236,6 +1244,8 @@
     t.boolean "cookies_acknowledged", default: false, null: false
     t.boolean "registration_notifications_enabled", default: false
     t.string "otp_secret"
+    t.integer "senior_delegate_id"
+    t.string "delegate_status", limit: 64
```
Comment (gregorbg) [inline db/schema.rb:1248]: Definitely unrelated

**Review by gregorbg — CHANGES_REQUESTED** [review]

File: app/controllers/api/v0/user_roles_controller.rb (hunk 5 above)
Comment (danieljames-dj) [reply app/controllers/api/v0/user_roles_controller.rb]: I've changed this a bit in https://github.com/thewca/worldcubeassociation.org/pull/9378 and now this method is living in user_role.rb with the name `can_user_read?`, this should make it more clear.

This method is only used for read API. The reason why `can_edit_groups` is changed to `can_read_groups` is that, earlier both were having same value - ie, a user can edit a group if they can read the hidden group. But for banned competitors it's different - eg: WEC have access to read the hidden group, but not write. So created new permission.

File: app/controllers/api/v0/user_roles_controller.rb (hunk 6 above)
Comment (danieljames-dj) [reply app/controllers/api/v0/user_roles_controller.rb:237]: I'll add them once https://github.com/thewca/worldcubeassociation.org/pull/9395 is merged since this PR is dependent on that 😬

File: app/models/user.rb (hunk 7 above)
Comment (danieljames-dj) [reply app/models/user.rb:586]: Because users like WEC members can view banned competitors, but they cannot edit banned competitors.

File: db/schema.rb (hunk 8 above)
Comment (danieljames-dj) [reply db/schema.rb]: Oops removed

File: db/schema.rb (hunk 9 above)
Comment (danieljames-dj) [reply db/schema.rb]: Oops removed

File: app/controllers/api/v0/user_roles_controller.rb (hunk 6 above)
Comment (gregorbg) [reply app/controllers/api/v0/user_roles_controller.rb:237]: Now that's merged, yay!

File: app/controllers/api/v0/user_roles_controller.rb (hunk 5 above)
Comment (gregorbg) [reply app/controllers/api/v0/user_roles_controller.rb]: Okay, makes sense. Thanks for the explanation!

File: app/models/user.rb (hunk 7 above)
Comment (gregorbg) [reply app/models/user.rb:586]: Okay, this makes sense but still looks a bit strange because one might be under the impression that you're giving Banned a kind of "special" treatment. I know this is due to the fact that they are currently the only group that could potentially have read-only access, but at least in the back of your mind think about something more generic for computing permissions, please.

File: app/controllers/api/v0/user_roles_controller.rb [hunk 10]
```diff
@@ -239,6 +244,29 @@ def update
       else
         return render status: :unprocessable_entity, json: { error: "Invalid parameter to be changed" }
       end
+    elsif group_type == UserGroup.group_types[:banned_competitors]
+      if params.key?(:endDate)
+        role.end_date = params.require(:endDate)
+      end
+
+      role.save!
```
Comment (danieljames-dj) [inline app/controllers/api/v0/user_roles_controller.rb:252]: role.update can be used, but there are more parameters to be added in the next PR, so updating each inside if block and then role.save at end is ideal here.

File: app/controllers/api/v0/user_roles_controller.rb [hunk 11]
```diff
@@ -239,6 +244,29 @@ def update
...
+            'End Date'
+          else
+            nil
+          end
+        )
+        if changed_parameter.present?
+          changes << UserRole::UserRoleChange.new(
+            changed_parameter: changed_parameter,
+            previous_value: values[0] || 'None',
+            new_value: values[1] || 'None',
```
Comment (danieljames-dj) [inline app/controllers/api/v0/user_roles_controller.rb:266]: @gregorbg Could you please review these two lines in more deep? I'm looking to print 'None' whenever the value is nil. This is working fine, but I'm confused whether it will miss any edge cases. Like, I expected `0 || 'None'` to give 'None', but it gives 0. Luckily I wanted 0. (In python it would have taken 'None') Basically I want to take 'None' only if value is `nil`.

Comment (gregorbg) [reply app/controllers/api/v0/user_roles_controller.rb:268]: Yup, interpreted languages all have their own rules about what is false-y and what isn't.
Ruby is relatively strict in comparison, but also has a lot of convenience methods for you: `my_value.nil?` should do exactly what you want.

Comment (gregorbg) [reply app/controllers/api/v0/user_roles_controller.rb:268]: You can de-structure the array (similar to JS `const [foo, bar] = my_awesome_array` but slightly different syntax) using `previous_value, new_value = values`. Then you can have a helper function `changed_value_to_human_readable`:

```ruby
def changed_value_to_human_readable(value)
  value.nil? ? 'None' : value
end
```

Comment (gregorbg) [reply app/controllers/api/v0/user_roles_controller.rb:268]: Using the above method and destructuring, you can then do 
```suggestion
            previous_value: changed_value_to_human_readable(previous_value),
            new_value: changed_value_to_human_readable(new_value),
```
and I think that's reasonably readable

**Review by gregorbg — COMMENTED** [review]
Code is okay, I've given a suggestion about the snippet that you asked for.

Why are there no tests changing in this PR at all? Banned Competitors is a quite sensitive area of our code, I'd like this backend change to be backed and supported by our existing tests.

File: app/controllers/api/v0/user_roles_controller.rb (hunk 11 above)
Comment (danieljames-dj) [reply app/controllers/api/v0/user_roles_controller.rb]: Thanks, I've implemented your suggestion.

**Review by danieljames-dj — COMMENTED** [review]
Thanks Gregor, regarding unit tests - I was planning to have the unit tests for all methods in user_controller as another PR. But since you explicitly mentioned, I've added it for the read APIs. And while writing that, I realized a mistake during rebase - I didn't change `can_write_group` to `can_read_group` because of which banned competitors didn't become accessible to delegates.

File: spec/controllers/api/v0/user_roles_controller_spec.rb [hunk 12]
```diff
@@ -28,6 +29,28 @@
...
+    context 'when user is logged in as a normal user' do
+      sign_in { FactoryBot.create(:user) }
+
+      it 'fetches list of roles of a user' do
+        get :index_for_user, params: { user_id: user_whose_delegate_status_changes.id }
+
+        expect(response.body).to eq(user_whose_delegate_status_changes.active_roles.to_json)
+      end
+
+      it 'fetches list of banned competitos' do
```
Comment (gregorbg) [inline spec/controllers/api/v0/user_roles_controller_spec.rb:49]: The point of this test is that it does NOT fetch the list of banned competitors :P

**Review by gregorbg — APPROVED** [review]
Small nitpick about the naming of the tests, but the rest of the code is okay so I'll approve already and you can merge once you fixed the nitpick

File: spec/controllers/api/v0/user_roles_controller_spec.rb (hunk 12 above)
Comment (danieljames-dj) [reply spec/controllers/api/v0/user_roles_controller_spec.rb:49]: Thanks, changed. 👍
